File: android/app/src/main/java/com/typeeasy/KeyboardSuggestionLexicon.kt
package com.typeeasy

/**
 * On-device suggestions when Datamuse misses fuzzy prefixes (e.g. `taml` → tamil).
 */
object KeyboardSuggestionLexicon {
    private val words: List<String> = buildSet {
        add("tamil")
        add("tamilnadu")
        add("tamil nadu")
        add("chennai")
        add("madurai")
        add("coimbatore")
        add("translate")
        add("translation")
        add("grammar")
        add("voice")
        add("keyboard")
        add("settings")
        add("hello")
        add("thanks")
        add("please")
        add("india")
        add("english")
        add("hindi")
        add("telugu")
        add("malayalam")
        add("kannada")
        add("marathi")
        add("bengali")
        listOf(
            "en" to "English", "ta" to "Tamil", "hi" to "Hindi", "fr" to "French",
            "de" to "German", "es" to "Spanish", "ar" to "Arabic", "zh" to "Chinese",
            "ja" to "Japanese", "ko" to "Korean", "ru" to "Russian", "pt" to "Portuguese",
            "it" to "Italian", "bn" to "Bengali", "te" to "Telugu", "ml" to "Malayalam",
            "kn" to "Kannada", "mr" to "Marathi",
        ).forEach { (code, name) ->
            add(code)
            add(name)
            add(name.lowercase())
        }
    }.toList().sorted()

    fun suggestions(query: String, max: Int = 8): List<String> {
        val q = query.trim().lowercase()
        if (q.isEmpty()) return emptyList()
        val scored = words.mapNotNull { word ->
            val w = word.lowercase()
            matchScore(q, w)?.let { score -> word to score }
        }.sortedWith(compareBy<Pair<String, Int>> { it.second }.thenBy { it.first.length }.thenBy { it.first })

        val seen = mutableSetOf<String>()
        val out = ArrayList<String>(max)
        for ((word, _) in scored) {
            val key = word.lowercase()
            if (!seen.add(key)) continue
            out.add(displayForm(word, q))
            if (out.size >= max) break
        }
        return out
    }

    private fun matchScore(query: String, word: String): Int? =
        KeyboardSuggestionMatcher.matchScore(query, word)

    private fun displayForm(word: String, query: String): String =
        KeyboardSuggestionMatcher.displayForm(word, query)
}

File: ios/Shared/KeyboardSuggestionLexicon.swift
import Foundation

/// On-device word list + fuzzy prefix match (Datamuse alone misses e.g. `taml` → tamil).
enum KeyboardSuggestionLexicon {
  private static let words: [String] = {
    var set = Set<String>()
    func add(_ w: String) { set.insert(w.lowercased()) }

    for lang in KeyboardLanguages.all {
      add(lang.name)
      add(lang.code)
    }

    let extra = [
      "tamil", "tamilnadu", "tamil nadu", "chennai", "madurai", "coimbatore",
      "english", "hindi", "telugu", "malayalam", "kannada", "marathi", "bengali",
      "translate", "translation", "grammar", "voice", "keyboard", "settings",
      "hello", "thanks", "please", "sorry", "welcome", "good", "morning", "evening",
      "india", "america", "london", "paris", "tokyo", "beijing",
      "message", "email", "phone", "call", "meeting", "today", "tomorrow", "yesterday",
    ]
    extra.forEach { add($0) }
    return Array(set).sorted()
  }()

  static func suggestions(for query: String, max: Int = 8) -> [String] {
    let q = query.lowercased().trimmingCharacters(in: .whitespacesAndNewlines)
    guard q.count >= 1 else { return [] }

    var scored: [(word: String, score: Int)] = []
    for word in words {
      let w = word.lowercased()
      if let score = KeyboardSuggestionMatcher.matchScore(query: q, word: w) {
        scored.append((word, score))
      }
    }

    scored.sort { a, b in
      if a.score != b.score { return a.score < b.score }
      if a.word.count != b.word.count { return a.word.count < b.word.count }
      return a.word < b.word
    }

    var seen = Set<String>()
    var out: [String] = []
    for item in scored {
      if seen.contains(item.word) { continue }
      seen.insert(item.word)
      out.append(KeyboardSuggestionMatcher.displayForm(item.word, query: q))
      if out.count >= max { break }
    }
    return out
  }
}
